sample-compose/data/src/main/java/co/nimblehq/sample/compose/data/repository/AppPreferencesRepositoryImpl.kt

```diff
@@ -0,0 +1,44 @@
+package co.nimblehq.sample.compose.data.repository
+
+import android.util.Log
+import androidx.datastore.core.DataStore
+import androidx.datastore.preferences.core.*
+import co.nimblehq.sample.compose.domain.repository.AppPreferencesRepository
+import kotlinx.coroutines.flow.*
+import java.io.IOException
+import javax.inject.Inject
+
+class AppPreferencesRepositoryImpl @Inject constructor(
+    private val appPreferencesDataStore: DataStore<Preferences>
+) : AppPreferencesRepository {
+
+    private object PreferencesKeys {
+        val FIRST_TIME_LAUNCH = booleanPreferencesKey("FIRST_TIME_LAUNCH")
+    }
+
+    override fun isFirstTimeLaunchPreferences(): Flow<Boolean> = flow {
+        val firstLaunchPreferences = appPreferencesDataStore.data
+            .catch { exception ->
+                if (exception is IOException) {
+                    Log.e(
+                        AppPreferencesRepositoryImpl::class.simpleName,
+                        "Error reading preferences.",
+                        exception
+                    )
+                    emit(emptyPreferences())
+                } else {
+                    throw exception
+                }
+            }.map { preferences ->
+                preferences[PreferencesKeys.FIRST_TIME_LAUNCH] ?: true
+            }.first()
+
+        emit(firstLaunchPreferences)
+    }
```

Comment:
I'm not sure why do we need to wrap the result of `appPreferencesDataStore.data` inside a `flow {}` block again with `.first` to get the value since its result is a `Flow<Boolean>` already. Why don't we just pass the result to the main fun like this?

```suggestion
    override fun isFirstTimeLaunchPreferences(): Flow<Boolean> = appPreferencesDataStore.data
        .catch { exception ->
            if (exception is IOException) {
                Log.e(
                    AppPreferencesRepositoryImpl::class.simpleName,
                    "Error reading preferences.",
                    exception
                )
                emit(emptyPreferences())
            } else {
                throw exception
            }
        }.map { preferences ->
            preferences[FIRST_TIME_LAUNCH] ?: true
        }
```

Reply:
@luongvo At first, I implemented as your suggestion. But I found that the receiver will never go into the collect() function and keeps waiting for the next value.

Reply:
@lydiasama That's the benefit of `Flow` lol to observe value changes in ViewModel, like we can observe value changes in the Room database with LivedData. Please look at the [first sample code of this question in stackoverflow](https://stackoverflow.com/questions/70567126/wait-for-preferences-datastore-to-retrieve-data-from-preferences-datastore-in-wo), they use it as an observer. So:
- We should not put Loading state with `isFirstTimeLaunchPreferencesUseCase()` execution.
- When ever new value comes, it will delegate the new value to the UI layer. If we don't want to keep the observer, use `first()`, but in VM layer instead.

```
        execute {
            showLoading()
            getModelsUseCase()
                .catch {
                    _error.emit(it)
                }
                .collect { result ->
                    val uiModels = result.toUiModels()
                    _uiModels.emit(uiModels)
                }
            hideLoading()
        }
        execute {
            // Use first() instead of collect() if we don't want to observe the value changes from DataStore  
            isFirstTimeLaunchPreferencesUseCase()
                .catch {
                    _error.emit(it)
                }
                .collect { isFirstTimeLaunch ->
                    _isFirstTimeLaunch.emit(isFirstTimeLaunch)
                    if (isFirstTimeLaunch) {
                        updateFirstTimeLaunchPreferencesUseCase(false)
                    }
                }
        }
```

Reply:
@luongvo I see. Then I'll move `.first()` at the use case level. Thanks.

Reply:
@lydiasama I think we should use `first()` at ViewModel layer because `first()` will return the raw boolean value instead of Flow.